File: backend/trie/trie.go
package trie 

// create a trie data structure
type Trie struct {
	root *Node
}

// create a node data structure
type Node struct {
	children map[rune]*Node
	isEnd    bool
}

// create a new trie
func NewTrie() *Trie {
	return &Trie{
		root: NewNode(),
	}
}

// create a new node
func NewNode() *Node {
	return &Node{
		children: make(map[rune]*Node),
	}
}

// insert a word into the trie
func (t *Trie) Insert(word string) {
	node := t.root
	for _, char := range word {
		if _, ok := node.children[char]; !ok {
			node.children[char] = NewNode()
		}
		node = node.children[char]
	}
	node.isEnd = true
}

// search a word in the trie
func (t *Trie) Search(word string) bool {
	node := t.root
	for _, char := range word {
		if _, ok := node.children[char]; !ok {
			return false
		}
		node = node.children[char]
	}
	return node.isEnd
}

// starts with prefix
func (t *Trie) StartsWith(prefix string) bool {
	node := t.root
	for _, char := range prefix {
		if _, ok := node.children[char]; !ok {
			return false
		}
		node = node.children[char]
	}
	return true
}

// get all words in the trie
func (t *Trie) GetAllWords() []string {
	words := []string{}
	t.getWords(t.root, "", &words)
	return words
}

// get words recursively
func (t *Trie) getWords(node *Node, word string, words *[]string) {
	if node.isEnd {
		*words = append(*words, word)
	}
	for char, childNode := range node.children {
		t.getWords(childNode, word+string(char), words)
	}
}

// delete a word from the trie
func (t *Trie) Delete(word string) {
	t.delete(t.root, word, 0)
}

// delete recursively
func (t *Trie) delete(node *Node, word string, index int) {
	if index == len(word) {
		node.isEnd = false
		return
	}
	char := rune(word[index])
	if _, ok := node.children[char]; !ok {
		return
	}
	t.delete(node.children[char], word, index+1)
	if len(node.children[char].children) == 0 {
		delete(node.children, char)
	}
}

// get all words with a given prefix
func (t *Trie) GetWordsWithPrefix(prefix string) []string {
	words := []string{}
	node := t.root
	for _, char := range prefix {
		if _, ok := node.children[char]; !ok {
			return words
		}
		node = node.children[char]
	}
	t.getWords(node, prefix, &words)
	return words
}

// get k words with a given prefix
func (t *Trie) GetKWordsWithPrefix(prefix string, k int) []string {
	words := []string{}
	node := t.root
	for _, char := range prefix {
		if _, ok := node.children[char]; !ok {
			return words
		}
		node = node.children[char]
	}
	t.getKWords(node, prefix, &words, k)
	return words
}

// get k words recursively
func (t *Trie) getKWords(node *Node, word string, words *[]string, k int) {
	if node.isEnd {
		*words = append(*words, word)
	}
	if len(*words) == k {
		return
	}
	for char, childNode := range node.children {
		t.getKWords(childNode, word+string(char), words, k)
		if len(*words) == k {
			return
		}
	}
}

File: client/src/trie/trie.ts
// a classic trie implementation which consumes strings

// trie node class
class TrieNode {
  char: string;
  children: { [key: string]: TrieNode };
  is_end: boolean;
  constructor(char: string) {
    this.char = char;
    this.children = {};
    this.is_end = false;
  }
}

export class Trie {
  root: TrieNode;
  constructor() {
    this.root = new TrieNode("");
  }

  insert(word: string) {
    let curr: TrieNode = this.root;
    for (let i = 0; i < word.length; i++) {
      let char = word.charAt(i);
      if (!curr.children[char]) {
        curr.children[char] = new TrieNode(char);
      }
      curr = curr.children[char];
    }
    curr.is_end = true;
  }

  search(word: string): boolean {
    let curr: TrieNode = this.root;
    for (let i = 0; i < word.length; i++) {
      let char = word.charAt(i);
      if (!curr.children[char]) {
        return false;
      }
      curr = curr.children[char];
    }
    return curr.is_end;
  }

  startsWith(prefix: string): boolean {
    let curr: TrieNode = this.root;
    for (let i = 0; i < prefix.length; i++) {
      let char = prefix.charAt(i);
      if (!curr.children[char]) {
        return false;
      }
      curr = curr.children[char];
    }
    return true;
  }

  // returns k words in the trie that start with the given prefix
  // if there are less than k words all words are returned
  // if no words are found, returns an empty array
  getKWords(prefix: string, k: number): string[] {
    let curr: TrieNode = this.root;
    for (let i = 0; i < prefix.length; i++) {
      let char = prefix.charAt(i);
      if (!curr.children[char]) {
        return [];
      }
      curr = curr.children[char];
    }
    return this.getWordsHelper(curr, prefix, k);
  }

  // helper function for getWords
  getWordsHelper(node: TrieNode, prefix: string, k: number): string[] {
    let words: string[] = [];
    if (node.is_end) {
      words.push(prefix);
    }
    for (let child in node.children) {
      if (words.length >= k) {
        break;
      }
      words = words.concat(this.getWordsHelper(node.children[child], prefix + child, k-words.length));
    }
    return words;
  }
}